### DISCUSSION TITLE
[General] Gas price estimation on ZKsync chains

### DISCUSSION THREAD

**Discussion by daniel-savu** [zkSync-Community-Hub/zksync-developers#832] [General] [answered]
### Team or Project

Hyperlane

### Environment

Mainnet

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

Transaction submission

### Additional Details

Some questions:
1. What's the recommeneded way of estimating gas price when submitting to a zkSync-stack chain? 
    - Based on the response in https://github.com/zkSync-Community-Hub/zksync-developers/discussions/808#discussioncomment-11462416 it sounds like gas price should always be estimated using `zks_estimateFee` but I want to double-check?
1. Are eip-1559 parameters used at all or is it only the base fee that matters?
    - ZkSync docs [here](https://docs.zksync.io/zksync-protocol/rollup/transaction-lifecycle) suggest `maxFeePerGas` and `maxPriorityFeePerGas` aren't being used at all - is that still the case and will this ever change? Asking since `zks_estimateFee` doesn't return a value for `baseFee`, so wondering if there are any recommendations for estimating `baseFee`
1. Does the gas-per-pubdata parameter (`gas_per_pubdata_limit`) need be specified? Since this field is present in the eip-712 format but not in eip-1559, does sending txs as eip-712 provide better inclusion guarantees / compatiblity?

We have a Rust PR that attempts to follow the same gas estimation approach from the JS zkSync sdk, and it seems to work ok without a `gas_per_pubdata_limit` parameter so far (just using eip-1559 format). Any feedback there would be appreciated: https://github.com/hyperlane-xyz/hyperlane-monorepo/pull/5007

**Comment by alhenaai** [comment]
### Gas Price Estimation on ZKsync Chains

1. **Estimating Gas Price:**
   - Yes, the recommended way to estimate gas prices when submitting transactions to a ZKsync chain is indeed using the `zks_estimateFee` method. This method provides an estimate of the gas and fee details for a given call request, ensuring compatibility with the ZKsync Era's fee structure. More details can be found in the [ZKsync JSON-RPC API documentation](https://docs.zksync.io/zksync-protocol/api/zks-rpc).

2. **Use of EIP-1559 Parameters:**
   - In ZKsync Era, while the EIP-1559 transaction format is supported, the `maxFeePerGas` and `maxPriorityFeePerGas` parameters are not utilized. Currently, only the base fee is relevant in this context. There are no indications in the provided documentation about plans to change this implementation. For more details, refer to the [Transaction Lifecycle documentation](https://docs.zksync.io/zksync-protocol/rollup/transaction-lifecycle).

3. **Gas-per-Pubdata Parameter:**
   - The `gas_per_pubdata_limit` parameter is crucial in ZKsync-specific EIP712 transactions, as it indicates the maximum gas price the operator can charge users per byte of pubdata. This field is not present in EIP-1559 transactions but is essential in EIP-712 for managing costs related to pubdata. If your implementation works without explicitly setting this parameter, it might be because the SDK or the network defaults are handling it adequately. However, specifying it could provide better control over transaction costs related to pubdata.

For further details and discussions, you might find it helpful to visit the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by JackHamer09** [comment, marked as answer]
Hey! Here’s some clarification:

1. **Gas Price Estimation:**  
   Yes, it’s recommended to use `zks_estimateFee`. This works seamlessly with the `zksync-ethers` and `viem` SDKs.

2. **EIP-1559 Parameters:**  
   In fact `maxFeePerGas` is utilized and includes the base fee, so it’s sufficient to use that in transactions. `maxPriorityFeePerGas` is not utilized, as ZKsync doesn’t have a concept of priority fees. The server likely returns `0` for this parameter. We're currently fixing the documentation to make this clearer and fix that error.

3. **`gasPerPubdata` Parameter:**  
   This is only required for `eip-712` transactions. You can continue using the `eip-1559` format, which ZKsync supports well. Transaction inclusion shouldn’t differ between the two formats.

Hope this helps!
